internal/fs/buffer/inmemory_write_buffer_test.go

```diff
@@ -41,24 +56,371 @@ func (t *MemoryBufferTest) SetUp(ti *TestInfo) {
+	t.validateInMemoryBuffer(t.mb.flushed, 0, 0, 0, 0)
+}
+
+func (t *MemoryBufferTest) TestWriteRollOverFromCurrentToFlushedBuffer() {
+	t.createAndValidateInMemoryBuffer(bufferSizeInMB)
+	data1 := t.generateRandomData(MiB)
+	data2 := t.generateRandomData(2 * KiB)
+
+	// Write to buffer
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, 2*MiB-KiB)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, 2*MiB+KiB)
+	AssertTrue(bytes.Equal(data2[0:KiB], t.mb.flushed.buffer[MiB-KiB:MiB]))
+	AssertTrue(bytes.Equal(data2[KiB:], t.mb.current.buffer[0:KiB]))
+	t.validateInMemoryBuffer(t.mb.current, bufferSizeInMB*MiB, 0, 2*MiB, 3*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, bufferSizeInMB*MiB, 0, MiB, 2*MiB)
+}
+
+func (t *MemoryBufferTest) TestRandomWriteOnAnAlreadyUploadedBufferBlockShouldFail() {
+	t.createAndValidateInMemoryBuffer(bufferSizeInMB)
+	data1 := t.generateRandomData(MiB)
+	data2 := t.generateRandomData(2 * KiB)
+
+	// Write to buffer
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	// Write to an already uploaded offset
+	err = t.mb.WriteAt([]byte(smallContent3), 0)
+
+	AssertNe(nil, err)
+	AssertEq(NonSequentialWriteError, err.Error())
+	AssertEq(t.mb.fileSize, MiB+2*KiB)
+	t.validateInMemoryBuffer(t.mb.current, bufferSizeInMB*MiB, 0, MiB, 2*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, bufferSizeInMB*MiB, 0, 0, MiB)
+}
+
+func (t *MemoryBufferTest) TestRandomWriteBeyondCurrentBufferAfterSequentialWritesShouldFail() {
+	t.createAndValidateInMemoryBuffer(bufferSizeInMB)
+	data1 := t.generateRandomData(MiB)
+	data2 := t.generateRandomData(2 * KiB)
+
+	// Sequential writes
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	// Random write beyond current buffer block
+	err = t.mb.WriteAt(data2, 2*MiB+1)
+
+	AssertNe(nil, err)
+	AssertEq(NonSequentialWriteError, err.Error())
+	AssertEq(t.mb.fileSize, MiB+2*KiB)
+	t.validateInMemoryBuffer(t.mb.current, bufferSizeInMB*MiB, 0, MiB, 2*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, bufferSizeInMB*MiB, 0, 0, MiB)
+}
+
+func (t *MemoryBufferTest) TestSmallContentWriteWith2ChunksSkippedShouldFail() {
+	t.createAndValidateInMemoryBuffer(bufferSizeInMB)
+	data1 := t.generateRandomData(MiB)
+	data2 := t.generateRandomData(2 * KiB)
+
+	// Write to buffer sequentially
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	// Write to an offset with 2 chunks skipped
+	err = t.mb.WriteAt([]byte(smallContent1), 5*MiB)
+
+	AssertNe(nil, err)
+	AssertEq(NonSequentialWriteError, err.Error())
+	AssertEq(t.mb.fileSize, MiB+2*KiB)
+	t.validateInMemoryBuffer(t.mb.current, bufferSizeInMB*MiB, 0, MiB, 2*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, bufferSizeInMB*MiB, 0, 0, MiB)
+}
+
+func (t *MemoryBufferTest) TestBigContentWriteWith2ChunksSkippedShouldFail() {
```

Comment:
Another useful test would when the `contentSize > bufferSize` and the final bits of content go even beyond the next current buffer (i.e. the content doesn't fit in the next current buffer even after advancing), because we don't advance more than once in the code today.
I know it's not a problem in real use case because kernel will never send more than 1 MB (which is <= bufferSize), but let's a negative test to confirm it.